subsys/bluetooth/host/hci_core.c

```diff
@@ -1078,12 +1105,61 @@ static struct bt_conn *find_pending_connect(bt_addr_le_t *peer_addr)
 	return bt_conn_lookup_state_le(peer_addr, BT_CONN_CONNECT_DIR_ADV);
 }
 
+static void conn_auto_initiate(struct bt_conn *conn)
+{
+	int err;
+
+	if (conn->state != BT_CONN_CONNECTED) {
+		/* It is possible that connection was disconnected directly from
+		 * connected callback so we must check state before doing
+		 * connection parameters update.
+		 */
+		return;
+	}
+
+	if (!atomic_test_bit(conn->flags, BT_CONN_AUTO_FEATURE_EXCH) &&
+	    ((conn->role == BT_HCI_ROLE_MASTER) ||
+	     BT_FEAT_LE_SLAVE_FEATURE_XCHG(bt_dev.le.features))) {
+		err = hci_le_read_remote_features(conn);
+		if (!err) {
```

Comment:
Hi guys,

I wonder whether the condition is correct. Shouldn't it be just `if (err)` ? It looks like it was borrowed from the commits this PR was based on.

I found this issue when tried to update connection parameters from the app level. They are simply not updated, because of the timer-specifics involved in the call of slave_update_conn_param() which is below in this function. And this `slave_update_conn_param()` is not called when conditions like I mentioned return from the `conn_auto_initiate()`..

Reply:
`if (!err)` is the correct expression here. It returns on success off sending the command. If not success it moves to the next one. You can see it goes for all of the HCI procedures in this function. Once the procedure completes this function will be called and the next procedure should be initiated.

Reply:
@dnsglk The intention is to call all of them. But the three first in the list  are asynchronous, and only one can be active at a time.
If `hci_le_read_remote_features` is called, then `conn_auto_initiate` will be called again from `bt_hci_evt_read_remote_version_complete`, and this time it will call the next function in your list. 
`hci_le_set_data_len()` and `slave_update_conn_param()` can be called simultaneously, so they don't have this behavior.

Reply:
I see now. In my case then, gdb shows that `le_phy_update_complete()` is not called for whatever reason...

Reply:
@joerchan thanks for spending time explaining it. Anyway, it still looks like something to improve IMO. Because, if remote doesn't respond to that query (which is my guess on why the phy callback is not triggered), it still means that not all of auto initialization has been completed.

Reply:
@dnsglk Can you create an bug-report with the missing `le_phy_update_complete` and tell us how to reproduce this problem? Please be as detailed as possible

Reply:
@joerchan  I'll post a report if it I have more time to dig deeper into the issue. Thanks for understanding.